### PR TITLE
Andy/jsonbagg

### PR DESCRIPTION
**Pull request by andrew-wm-arthur**
This PR continues the generalization of Nautobot towards being MySQL compatible. 

- Removes Postgres specific collations `CollateAsChar`
- Replaces `django.contrib.postgres.aggregates.JSONBAgg` with a custom JSON aggregation function
- Removes aggregation from `ConfigContextQuerySet.get_for_object()` by using `.values_list()` in `ConfigContextModel.get_config_context()`. This is necessary because MySQL does not support ordering within JSON aggregation functions.

### CODE REVIEW

File: nautobot/extras/querysets.py [hunk 1]
```diff
@@ -70,14 +65,17 @@ class ConfigContextModelQuerySet(RestrictedQuerySet):
 
     def annotate_config_context_data(self):
```
Comment (andrew-wm-arthur) [inline nautobot/extras/querysets.py:66]: Given the changes to `ConfigContextModel.get_config_context()`, `annotate_config_context_data()` is now only used in views. However, we may still need to iterate here to completely remove the dependency ordered JSON aggregation.

File: nautobot/utilities/query_functions.py [hunk 2]
```diff
@@ -30,3 +35,11 @@ class EmptyGroupByJSONBAgg(OrderableJSONBAgg):
     """
 
     contains_aggregate = False
+
+
+class OrderableJSONBAgg(OrderableAggMixin, JSONBAgg):
+    """
+    TODO in Django 3.2 ordering is supported natively on JSONBAgg so this is no longer needed.
```
Comment (jathanism) [inline nautobot/utilities/query_functions.py:42]: Hmm... Perhaps we should go ahead and upgrade to Django 3.2 and obviate this?

File: nautobot/dcim/tests/test_natural_ordering.py [hunk 3]
```diff
@@ -34,35 +34,35 @@ def test_interface_ordering_numeric(self):
...
             "0:1.2",
             "0:1.10",
+            "0.2",
             "0:2",
             "0:2.0",
             "0:2.1",
             "0:2.2",
             "0:2.10",
+            "0.10",
+            "0.100",
```
Comment (jathanism) [inline nautobot/dcim/tests/test_natural_ordering.py:49]: At first glance this seems bad since is apparently breaking the natural ordering asserted by `CollateAsChar`... but then when you see the minor changes it actually looks more like a bug fix. 1.10 certainly should be coming after 1.2 for example!

File: nautobot/utilities/query_functions.py [hunk 4]
```diff
@@ -1,26 +1,31 @@
...
 
+class JSONBAgg(Aggregate):
+    function = None
+    output_field = JSONField()
 
-class OrderableJSONBAgg(OrderableAggMixin, JSONBAgg):
-    """
-    TODO in Django 3.2 ordering is supported natively on JSONBAgg so this is no longer needed.
-    """
+    # borrowed from django.contrib.postrgres.aggregats.JSONBagg
```
Comment (jathanism) [inline nautobot/utilities/query_functions.py:10]:
```suggestion
    # Borrowed from `django.contrib.postrgres.aggregates.JSONBagg`.
```

**Review by jathanism — APPROVED** [review]
LGTM. No major concerns.

File: nautobot/dcim/tests/test_natural_ordering.py (hunk 3 above)
Comment (jathanism) [reply nautobot/dcim/tests/test_natural_ordering.py:49]: @lampwins Can you please take a look-see at this one?

**Review by jathanism — APPROVED** [review]

File: nautobot/dcim/tests/test_natural_ordering.py (hunk 3 above)
Comment (lampwins) [reply nautobot/dcim/tests/test_natural_ordering.py:49]: I am not following this one. This looks like maybe an oversight in `.` vs `:`? I.e The original ordering has `0.2` before `0:1` which would be correct.

File: nautobot/virtualization/migrations/0001_initial.py [hunk 5]
```diff
@@ -383,7 +383,7 @@ class Migration(migrations.Migration):
                 "verbose_name": "interface",
                 "ordering": (
                     "virtual_machine",
-                    nautobot.utilities.query_functions.CollateAsChar("_name"),
+                    "_name",
```
Comment (lampwins) [inline nautobot/virtualization/migrations/0001_initial.py:386]: What is the reason for this change?

File: nautobot/virtualization/models.py [hunk 6]
```diff
@@ -424,7 +423,7 @@ class VMInterface(BaseModel, BaseInterface, CustomFieldModel):
 
     class Meta:
         verbose_name = "interface"
-        ordering = ("virtual_machine", CollateAsChar("_name"))
+        ordering = ("virtual_machine", "_name")
```
Comment (lampwins) [inline nautobot/virtualization/models.py:426]: Same ask here.

File: nautobot/dcim/migrations/0001_initial_part_1.py [hunk 7]
```diff
@@ -641,7 +641,7 @@ class Migration(migrations.Migration):
             options={
                 "ordering": (
                     "device",
-                    nautobot.utilities.query_functions.CollateAsChar("_name"),
+                    "_name",
```
Comment (lampwins) [inline nautobot/dcim/migrations/0001_initial_part_1.py:644]: Same ask here.

File: nautobot/dcim/models/device_components.py [hunk 8]
```diff
@@ -552,7 +551,7 @@ class Interface(CableTermination, PathEndpoint, ComponentModel, BaseInterface):
     ]
 
     class Meta:
-        ordering = ("device", CollateAsChar("_name"))
+        ordering = ("device", "_name")
```
Comment (lampwins) [inline nautobot/dcim/models/device_components.py:554]: Same ask here.

File: nautobot/extras/querysets.py (hunk 1 above)
Comment (lampwins) [reply nautobot/extras/querysets.py:66]: The reason for this is because in the device & VM views we need the actual config context objects to show other data about each object like its name and weight. Maybe if we move the deepmerge logic around we could get rid of this altogether and use the one method for both use cases?

**Review by lampwins — COMMENTED** [review]
The good news is the fundamental change looks sound and passes my manual testbed that was used for the original implementation upstream.

Otherwise, a couple of comments and questions.

File: nautobot/dcim/tests/test_natural_ordering.py (hunk 3 above)
Comment (andrew-wm-arthur) [reply nautobot/dcim/tests/test_natural_ordering.py:49]: I'm not clear on what the semantics are supposed to be here. Consider the case of:
`192.1, 192:1, 193.1, 193:1` with CollateAsChar. 

I proposed we change this to all IPv4 in numeric order, then all IPv6 in numeric order

File: nautobot/virtualization/migrations/0001_initial.py (hunk 5 above)
Comment (jathanism) [reply nautobot/virtualization/migrations/0001_initial.py:386]: I can attest that the reason for this is and other `_name` fields is eliminating `CollateasChar` since it's hard-coded to PG as a backend using the `COLLATE C` function. This doesn't exist in MySQL.

File: nautobot/dcim/tests/test_natural_ordering.py (hunk 3 above)
Comment (glennmatthews) [reply nautobot/dcim/tests/test_natural_ordering.py:49]: This is ordering of interfaces, subinterfaces, channelized interfaces, etc., not address ordering. Consider:

Ethernet1/1  (physical interface)
Ethernet1/1.100 (subinterface 100 of Ethernet1/1)
Ethernet1/1:2 (channelized or virtual interface 2 on Ethernet1/1)

From a UI standpoint it would be preferable for all `x:yyy` values to come either before or after all `x.zzz` values since they have different semantic meanings, rather than interleaving as `x.1`, `x:1`, `x.2`, `x:2`.

Comment (glennmatthews) [reply nautobot/dcim/tests/test_natural_ordering.py:49]: FWIW this goes all the way back to https://github.com/netbox-community/netbox/issues/9 (!) Though https://github.com/netbox-community/netbox/issues/2165 and https://github.com/netbox-community/netbox/issues/3097 are probably more relevant.

**Comment by andrew-wm-arthur** [comment]
@glennmatthews Thanks for the extra context on `CollateAsChar`. I thought it would be a straightforward change, so I lumped it in with the JSON aggregation changes, but that's not the case. 

I've reverted the collations-related changes and opened a separate PR: https://github.com/nautobot/nautobot/pull/289

**Comment by jathanism** [comment]
With the revered `CollateAsChar` changes, there's still the lingering issue of correctness in the `JSONBAgg` stuff, but I think we still need a final call from @lampwins on that.

**Review by lampwins — APPROVED** [review]
LGTM
